analyses/cms-open-data-ttbar/helper.cpp

```diff
@@ -31,10 +31,8 @@ TH1D Slice(TH1D &h, double low_edge, double high_edge)
 }
 
 // functions creating systematic variations
-inline TRandom &get_thread_local_trandom()
-{
-   thread_local TRandom rng;
-   rng.SetSeed(gRandom->Integer(1000));
+static TRandom &get_thread_local_trandom() {
+   static thread_local TRandom rng{gRandom->Integer(1000)};
```

Comment:
now that I look at this code, I don't think `gRandom->Integer` is safe to call concurrently from multiple threads like we do (and did)?

Reply:
I'm not really sure about that. I'm also not completely sure about why we do it, I just kept it there to avoid bringing unexpected changes. If that is really an issue, I see the following options:
1. Using the default constructor
2. Using a thread-safe rng to initialize the seed of TRandom (or not using TRandom altogether)
3. Locking the call to this function or calling `ROOT::EnableThreadSafety` inside of it.

Reply:
Number 5 was the original implementation. See the implementation of a class with N RNGs [here](https://github.com/andriiknu/analysis-grand-challenge/blob/b1e7dc8344bb03041dcbe7b11c39a5d6b20c655d/analyses/cms-open-data-ttbar/helper.cpp#L34). Then in order to use it we needed a separate call to
```
ROOT.gInterpreter.Declare(f"auto pt_res_up_obj = pt_res_up({ROOT.GetThreadPoolSize()});")
```
which imho it's not that nice (we would need to add it to the initialization function in both local RDF and DistRDF).
But what I found especially weird was the usage in `Vary` of all places
```
d = d.Vary('jet_pt', "ROOT::RVec<ROOT::RVecF>{jet_pt*pt_scale_up(), jet_pt*pt_res_up_obj(jet_pt, rdfslot_)}", ["pt_scale_up", "pt_res_up"])
```
This can be found in https://github.com/andriiknu/analysis-grand-challenge/blob/agc%2Brdf-2022/analyses/cms-open-data-ttbar/rdf_ttbar.ipynb

So no `DefineSlot` was used originally. And fortunately, as that wouldn't work in distributed mode (and there's no reason it should work)

Reply:
So I purposely reworked the generation of random gaussian numbers to avoid those weird patterns. In my latest commit I went for option 2, using only classes from the `std` and applying `thread_local` to every step. This should give a thread-safe generation IIUC. Let me know what you think

Reply:
> that wouldn't work in distributed mode

I'm not sure what "that" refers to exactly?

Reply:
`DefineSlot`